### DISCUSSION TITLE
Restful API URL Convention 제안

### DISCUSSION THREAD

**Discussion by YangTaeyoung** [project-555/plog-back#123] [General]
현재 전반적으로 API URL이 모두 Restful 하지 않아서 일반적으로 정의하는 URL의 룰을 정의하고자 합니다.
전반적으로 해당 개념이 부족할 때 작성된 것들이 있어 해당 부분을 점진적으로 수정해나갔으면 좋겠습니다. (이 부분은 프론트와 협의도 필요하겠지요)

# Resource CRUD API
- 자원을 생성, 조회, 수정, 삭제 하는 일반적인 액션을 수행하는 API의 경우 다음과 같이 보통 설계합니다.
  - `http://example.com/api/{version: ex. "v2"}/{domainName}/{resourceName}/{resourceKey}`

예를 들어 직장인의 자동차를 조회하는 API를 구성한다고 생각해봅시다.

직장인은 자동차를 여러대 살 수 있다고 가정하고 그 중 한 대만 조회합니다. 
각각의 직장인을 구분하는 키는 정수 형태인 직원 번호이며, 각각의 자동차를 구분하는 키는 정수 형태인 자동차의 번호라고 생각해봅시다.

예를 들어 5번 직장인의 자동차 리스트를 조회하는 API는 다음과 같이 구성할 수 있습니다.

## 다건
```http
GET http://example.com/api/v3/employees/5/cars

{
  "cars": [
    {
      "carID": 1,
      "name": "포르쉐",
      "color": "빨강"
    },
    {
      "carID": 2,
      "name": "람보르기니",
      "color": "파랑"
    },
    {
      "carID": 3,
      "name": "부가티베이론",
      "color": "노랑"
    },
  ]
}
```

## 단건 
5번 직장인의 3번 차, 부가티 베이론을 조회한다고 생각해보죠.
```http 
GET http://example.com/api/v3/employees/5/cars/3

{
  "car": {
      "carID": 3,
      "name": "부가티베이론",
      "color": "노랑"
    },
}
```

이처럼 직장인이라는 자원, 자동차라는 직작인의 서브 자원을 다음과 같이 자원을 나누어 조회하거나, 수정 할 수 있습니다.

이 외에도 상황에 맞게 POST, PUT, PATCH, DELETE와 같은 API 메서드를 곁들여 사용하면 좋겠죠.

## 기피해야 할 이름들
Restful API는 HTTP 메서드를 통해 자원에 대한 행위를 구분할 수 있을 만큼 주소에 생성, 수정, 삭제, 조회를 유추할 수 있는 네이밍의 경우에는 기피하는 것이 좋습니다.

```HTTP
GET /get-blogs # Bad
GET /blogs # Good

POST /blog-create # Bad
POST /blogs # Good

POST /blog-delete # Bad
DELETE /blogs/5 # Good

POST /update-blog # Bad 
PATCH /blogs/5 # Good
```

# Action API
자원과는 예외로 API 자체가 특정 비즈니스 로직을 어쩔 수 없이 품어야 하는 경우가 있습니다. 그 경우, 행위에 집중하여 API를 기술합니다.

예를 들어 로그인 API의 경우 토큰을 조회하는 API로 볼 수도 있지만, 아이디 및 비밀번호를 인증해야 하는 API로도 볼 수 있습니다. 이 경우 로그인 자체를 액션으로 두어 다음과 같이 작성할 수 있습니다.

```http
POST /auth/login # Good
{
  "accessToken": "1j2h6j12h6j1aa6",
  "expired": "2024-03-12T00:11:33+09:00"
}
```

**Comment by aiaaua** [comment]
일요일에 같이 이야기하고 백엔드부터 순차적으로 일괄 수정 후 프론트 반영하시죠!
정리하느라 고생많으셨습니다!
